crates/viewer/re_space_view_graph/src/types.rs

```diff
@@ -0,0 +1,79 @@
+use re_log_types::{EntityPath, Instance};
+use re_types::{components, ArrowString};
+
+use crate::graph::NodeIndex;
+
+impl<'a> EdgeInstance<'a> {
+    pub fn nodes(&'a self) -> impl Iterator<Item = &components::GraphNode> {
+        [&self.source, &self.target].into_iter()
+    }
+}
+
+impl<'a> From<&NodeInstance<'a>> for NodeIndex {
+    fn from(node: &NodeInstance<'a>) -> Self {
+        Self {
+            entity_hash: node.entity_path.hash(),
+            node_hash: node.node_id.into(),
+        }
+    }
+}
+
+impl<'a> From<NodeInstance<'a>> for NodeIndex {
+    fn from(node: NodeInstance<'a>) -> Self {
+        Self {
+            entity_hash: node.entity_path.hash(),
+            node_hash: node.node_id.into(),
+        }
+    }
+}
+
+pub struct NodeInstance<'a> {
+    pub node_id: &'a components::GraphNode,
+    pub entity_path: &'a EntityPath,
+    pub instance: Instance,
+    pub label: Option<&'a ArrowString>,
+    pub show_labels: bool,
+    pub color: Option<egui::Color32>,
+    pub position: Option<[f32; 2]>,
+}
+
+pub struct EdgeInstance<'a> {
+    pub source: components::GraphNode,
+    pub target: components::GraphNode,
+    pub entity_path: &'a re_log_types::EntityPath,
+    pub instance: Instance,
+    pub edge_type: components::GraphType,
+}
+
```

Comment:
We have a real opportunity to get rid of all these lifetimes by using owned values instead:
- `EntityPath` is cheap to clone (inner is an `Arc` + pre-computed hash)
- `ArrowString` and `GraphNode` are arrow buffer, which are also `Arc`